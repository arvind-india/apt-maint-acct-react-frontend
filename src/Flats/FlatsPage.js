import React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import { Table } from 'reactstrap'
import { Router } from 'react-router-dom'

import {
  MdAdd,
  MdVisibility,
  MdEdit,
  MdDelete
} from 'react-icons/lib/md' // material design icons

import {
  Button
} from 'reactstrap'

import { history } from '../_helpers'
import { PrivateRoute, FlashMessage } from '../_components'
import { flatActions as actions } from '../_actions'
import { FlatDetailsPage as detailsPage } from './FlatDetailsPage'

let url = '/flats'
let module = 'flats'

class FlatsPage extends React.Component {

  constructor(props) {
    super(props)
    this.handleDeleteModel = this.handleDeleteModel.bind(this)
  }

  componentDidMount() {
    this.props.dispatch(actions.getAll())
  }
  handleDeleteModel(id) {
    console.log('Deleting Flat with id: ', id)
    //return (e) => this.props.dispatch(actions.delete(id))
    this.props.dispatch(actions.delete(id))
    this.props.dispatch(actions.getAll()) // get list after deletion of a model
  }
  showList(models){
    const { authzn } = this.props

    let newModel = {
      model: {
        id: 0,
        block_number:'',
        flat_number: ''
      }
    }
    let addLink = authzn.allowsAdd ?
      <Link to={{ pathname: `${url}/0`, state: newModel }} title="Add"><MdAdd/></Link> :
      ''
    return <Table>
      <thead>
        <tr>
          <th>#</th>
          <th>Block Number</th>
          <th>Flat Number</th>
          <th>Actions {addLink}</th>
        </tr>
      </thead>
      <tbody>
        {models.items.map((model, index) =>
          <tr key={model.id}>
            <th scope="row">{index+1}</th>
            <td>{model.block_number}</td>
            <td>{model.flat_number}</td>
            <td>
              <Link
                to={{ pathname: `${url}/${model.id}`, state:{model: model} }}
                title={authzn.allowsEdit?"Edit":"View"}
              >{authzn.allowsEdit?<MdEdit/>:<MdVisibility/>}</Link>
              <Button
                color="link"
                title="Delete"
                onClick={() => this.handleDeleteModel(model.id)}
                hidden={!authzn.allowsDelete}
              ><MdDelete color="red"/></Button>
            </td>
          </tr>)}
      </tbody>
    </Table>
  }

  render() {
    console.log('Props in FlatsPage: ', this.props)
    const { flats, alert, authzn } = this.props
    let models = flats
    return (
      <div>
        <h3>Flats List</h3>
        {alert.message && <FlashMessage text={alert.message} delay={5000}/>}
        {models.loading && <em>Loading models...}</em>}
        {models.error && <span className="text-danger">ERROR: {models.error}</span>}
        {models.items && authzn && this.showList(models) }
        <Router history={history}>
          <div>
            <PrivateRoute path={`${url}/:id`} component={detailsPage} />
          </div>
        </Router>
      </div>
    )
  }
}

function mapStateToProps(state) {
  const { flats, alert, authorizations } = state
  // const { user } = authentication
  const authzn = authorizations[module]
  return {
//    user,
    flats,
    alert,
    authzn
  }
}

const connectedFlatsPage = connect(mapStateToProps)(FlatsPage)
export { connectedFlatsPage as FlatsPage }
