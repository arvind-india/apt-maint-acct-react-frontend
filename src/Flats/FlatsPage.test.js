// FlatsPage.test.js

import React from 'react'
import { shallow } from 'enzyme'

import { Flats } from './FlatsPage'

const mockProps = {
  trackHistory: false,  // for unit testing history is NOT required as it has 'keys' that are dynamic
  flats: {
    loading: false,
    error: false,
    items: [
      {
        id: 1,
        block_number: '0',
        flat_number: 'g1'
      },
      {
        id: 2,
        block_number: '0',
        flat_number: 'f1'
      },
      {
        id: 3,
        block_number: '1',
        flat_number: 'g4'      }
    ]
  },
  alert: {message: "testingAlert"},
  authzn: {
    allowsAdd: true,
    allowsEdit: true,
    allowsDelete: true,
    condition: ''
  },
  getAll: jest.fn(),
  delete: jest.fn()
}

const component = shallow(<Flats {...mockProps} />)

describe('FlatsPage params test', () => {
  it('should render FlatsPage without crashing', () => {
    component
  })
  it('should render initial layout', () => {
    expect(component.getElements()).toMatchSnapshot()
  })
})


describe('FlatsPage rows test', () => {
  let numberOfMockFlats = mockProps.flats.items.length
  let tbody = component.find('tbody')
  let rows = tbody.find('tr')
  it('should have '+numberOfMockFlats+' rows', () => {
    expect(rows.length).toEqual(numberOfMockFlats)
  })

})
