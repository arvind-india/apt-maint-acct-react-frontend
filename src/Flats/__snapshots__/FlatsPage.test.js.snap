// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlatsPage params test should render initial layout 1`] = `
Array [
  <div>
    <h3>
        Flats List
    </h3>
    <FlashMessage
        delay={5000}
        text="testingAlert"
    />
    <Table
        responsiveTag="div"
        tag="table"
    >
        <thead>
            <tr>
                <th>
                    #
                </th>
                <th>
                    Block Number
                </th>
                <th>
                    Flat Number
                </th>
                <th>
                    Actions 
                    <Link
                        replace={false}
                        title="Add"
                        to={
                            Object {
                                "pathname": "/flats/0",
                                "state": Object {
                                  "model": Object {
                                    "block_number": "",
                                    "flat_number": "",
                                    "id": 0,
                                  },
                                },
                              }
                        }
                    >
                        <MdAdd />
                    </Link>
                </th>
            </tr>
        </thead>
        <tbody>
            <tr>
                <th
                    scope="row"
                >
                    1
                </th>
                <td>
                    0
                </td>
                <td>
                    g1
                </td>
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/flats/1",
                                "state": Object {
                                  "model": Object {
                                    "block_number": "0",
                                    "flat_number": "g1",
                                    "id": 1,
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
            <tr>
                <th
                    scope="row"
                >
                    2
                </th>
                <td>
                    0
                </td>
                <td>
                    f1
                </td>
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/flats/2",
                                "state": Object {
                                  "model": Object {
                                    "block_number": "0",
                                    "flat_number": "f1",
                                    "id": 2,
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
            <tr>
                <th
                    scope="row"
                >
                    3
                </th>
                <td>
                    1
                </td>
                <td>
                    g4
                </td>
                <td>
                    <Link
                        replace={false}
                        title="Edit"
                        to={
                            Object {
                                "pathname": "/flats/3",
                                "state": Object {
                                  "model": Object {
                                    "block_number": "1",
                                    "flat_number": "g4",
                                    "id": 3,
                                  },
                                },
                              }
                        }
                    >
                        <MdEdit />
                    </Link>
                    <Button
                        color="link"
                        hidden={false}
                        onClick={[Function]}
                        tag="button"
                        title="Delete"
                    >
                        <MdDelete
                            color="red"
                        />
                    </Button>
                </td>
            </tr>
        </tbody>
    </Table>
    <Router
        history={Object {}}
    >
        <div>
            <PrivateRoute
                component={[Function]}
                path="/flats/:id"
            />
        </div>
    </Router>
</div>,
]
`;
